### PR TITLE
v4l2dec:use dma buf and add VIDIOC_ENUM_FMT ioctl

### PR DESCRIPTION
**Pull request by YuJiankang**
patch for VIZ-5168 and VIZ-5170

### CODE REVIEW

File: v4l2/v4l2_codecbase.cpp [hunk 1]
```diff
@@ -441,6 +441,16 @@ int32_t V4l2CodecBase::ioctl(int command, void* arg)
             DEBUG("%s port dqbuf->index: %d", THREAD_NAME(port), dqbuf->index);
         }
         break;
+        case VIDIOC_ENUM_FMT: {
+            struct v4l2_fmtdesc *fmtdesc = static_cast<struct v4l2_fmtdesc *>(arg);
+            if ((fmtdesc->index == 0) && (fmtdesc->type == V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE)) {
+                fmtdesc->pixelformat = V4L2_PIX_FMT_NV12M;
+            } else {
+                ret = -1;
+                break;
```
Comment (xuguangxin) [inline v4l2/v4l2_codecbase.cpp]: do not need this break

File: v4l2/v4l2_decode.cpp [hunk 2]
```diff
@@ -72,7 +72,7 @@ V4l2Decoder::V4l2Decoder()
     m_maxBufferSize[OUTPUT] = 0;
     m_bufferSpace[OUTPUT] = NULL;
 
-    m_memoryType = VIDEO_DATA_MEMORY_TYPE_DRM_NAME;
+    m_memoryType = VIDEO_DATA_MEMORY_TYPE_DMA_BUF;
```
Comment (xuguangxin) [inline v4l2/v4l2_decode.cpp]: better separate this to another commit

File: v4l2/v4l2_codecbase.cpp [hunk 3]
```diff
@@ -441,6 +441,16 @@ int32_t V4l2CodecBase::ioctl(int command, void* arg)
             DEBUG("%s port dqbuf->index: %d", THREAD_NAME(port), dqbuf->index);
         }
         break;
+        case VIDIOC_ENUM_FMT: {
```
Comment (xuguangxin) [inline v4l2/v4l2_codecbase.cpp]: is this decoder only? if it is please move this to v4l2_decoder

**Comment by YuJiankang** [comment]
separate to two commit and update with Guangxin's suggestion.

**Comment by xuguangxin** [comment]
+1

File: v4l2/v4l2_decode.cpp [hunk 4]
```diff
@@ -382,6 +382,15 @@ int32_t V4l2Decoder::ioctl(int command, void* arg)
...
     break;
+    case VIDIOC_ENUM_FMT: {
+        struct v4l2_fmtdesc *fmtdesc = static_cast<struct v4l2_fmtdesc *>(arg);
+        if ((fmtdesc->index == 0) && (fmtdesc->type == V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE)) {
+            fmtdesc->pixelformat = V4L2_PIX_FMT_NV12M;
+        } else {
+            ret = -1;
+        }
+    }
+    break;
```
Comment (chivakker) [inline v4l2/v4l2_decode.cpp]: This is a IOCTL used on both encoder and decoder.  It needs to be moved to the base class. I didn't see this review earlier.

Comment (xuguangxin) [inline v4l2/v4l2_decode.cpp]: ;) , sorry to close this too quickly.
Encoder need implement IOCTL by itself. The output format is different.We did not add encoder's IOCTL since we have no code to test it yet.
